modules/apps/29-fee/ibc_module.go

```diff
@@ -0,0 +1,253 @@
+	return im.app.OnChanOpenAck(ctx, portID, channelID, cpAppVersion)
+}
+
+// OnChanOpenConfirm implements the IBCModule interface
+func (im IBCModule) OnChanOpenConfirm(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	// call underlying app's OnChanOpenConfirm callback.
+	return im.app.OnChanOpenConfirm(ctx, portID, channelID)
+}
+
+// OnChanCloseInit implements the IBCModule interface
+func (im IBCModule) OnChanCloseInit(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	// delete fee enabled on channel
+	// and refund any remaining fees escrowed on channel
+	im.keeper.DeleteFeeEnabled(ctx, portID, channelID)
+	err := im.keeper.RefundFeesOnChannel(ctx, portID, channelID)
+	// error should only be non-nil if there is a bug in the code
+	// that causes module account to have insufficient funds to refund
+	// all escrowed fees on the channel.
+	// Disable all channels to allow for coordinated fix to the issue
+	// and mitigate/reverse damage.
+	// NOTE: Underlying application's packets will still go through, but
+	// fee module will be disabled for all channels
+	if err != nil {
+		im.keeper.DisableAllChannels(ctx)
+	}
+	return im.app.OnChanCloseInit(ctx, portID, channelID)
+}
+
+// OnChanCloseConfirm implements the IBCModule interface
+func (im IBCModule) OnChanCloseConfirm(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	// delete fee enabled on channel
+	// and refund any remaining fees escrowed on channel
+	im.keeper.DeleteFeeEnabled(ctx, portID, channelID)
+	err := im.keeper.RefundFeesOnChannel(ctx, portID, channelID)
+	// error should only be non-nil if there is a bug in the code
+	// that causes module account to have insufficient funds to refund
+	// all escrowed fees on the channel.
+	// Disable all channels to allow for coordinated fix to the issue
+	// and mitigate/reverse damage.
+	// NOTE: Underlying application's packets will still go through, but
+	// fee module will be disabled for all channels
+	if err != nil {
+		im.keeper.DisableAllChannels(ctx)
+	}
+	return im.app.OnChanCloseConfirm(ctx, portID, channelID)
+}
+
+// OnRecvPacket implements the IBCModule interface.
+// If fees are not enabled, this callback will default to the ibc-core packet callback
+func (im IBCModule) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) exported.Acknowledgement {
+	if !im.keeper.IsFeeEnabled(ctx, packet.DestinationPort, packet.DestinationChannel) {
+		return im.app.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	ack := im.app.OnRecvPacket(ctx, packet, relayer)
+
+	forwardRelayer, found := im.keeper.GetCounterpartyAddress(ctx, relayer.String())
+
+	// incase of async aknowledgement (ack == nil) store the ForwardRelayer address for use later
+	if ack == nil && found {
+		im.keeper.SetForwardRelayerAddress(ctx, channeltypes.NewPacketId(packet.GetSourceChannel(), packet.GetSourcePort(), packet.GetSequence()), forwardRelayer)
```

Comment:
The source channel and source port are not unique on the destination chain. This has to be set with the destination channel and port, otherwise it will get overridden when two source chains have the same source channel and port.

To make things clear, consider two separate functions `NewPacketIdForSourceChain(packet)` and `NewPacketIdForDestinationChain(packet)` to ensure that the packet ID is always derived uniquely on the respective chains.